### PR TITLE
feat(components): initial implementation of emulated content tag

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

File: modules/core/src/compiler/directive_metadata.js [hunk 1]
```diff
@@ -9,10 +10,13 @@ export class DirectiveMetadata {
   type:Type;
   annotation:Directive;
   shadowDomStrategy:ShadowDomStrategy;
+  componentDirectives:List<DirectiveMetadata>;
```
Comment (rkirov) [inline modules/core/src/compiler/directive_metadata.js]: componentDirectivesMetadata might be a more fitting name.

File: modules/core/src/compiler/directive_metadata_reader.js [hunk 2]
```diff
@@ -12,11 +13,17 @@ export class DirectiveMetadataReader {
...
+            type,
+            annotation,
+            shadowDomStrategy,
+            this.componentDirectives(annotation, shadowDomStrategy)
+          );
         }
 
         if (annotation instanceof Directive) {
-          return new DirectiveMetadata(type, annotation, null);
+          return new DirectiveMetadata(type, annotation, null, []);
```
Comment (rkirov) [inline modules/core/src/compiler/directive_metadata_reader.js]: shouldn't `componentDirectives = null` for a Directive (since it is not a component)?

File: modules/core/src/compiler/directive_metadata_reader.js [hunk 3]
```diff
@@ -26,4 +33,16 @@ export class DirectiveMetadataReader {
   parseShadowDomStrategy(annotation:Component):ShadowDomStrategy{
     return isPresent(annotation.shadowDom) ? annotation.shadowDom : ShadowDomNative;
   }
+
+  componentDirectives(annotation:Component, shadowDomStrategy:ShadowDomStrategy):List<DirectiveMetadata> {
+    var p = shadowDomStrategy.polyfillDirectives();
```
Comment (rkirov) [inline modules/core/src/compiler/directive_metadata_reader.js]: I am not sure about one letter local vars, can we pick something more descriptive like `polyDirs', `templateDirs`, they can still be short by abbreviating.

File: modules/core/src/compiler/directive_metadata_reader.js [hunk 4]
```diff
@@ -26,4 +33,16 @@ export class DirectiveMetadataReader {
...
+    var t = annotation.template;
+    var b = isPresent(t) && isPresent(t.directives) ? t.directives : [];
+
+    var res = [];
+    res = ListWrapper.concat(res, b)
```
Comment (rkirov) [inline modules/core/src/compiler/directive_metadata_reader.js]: Why not just reuse `b`, since polyfillDirectives returns a new array instance.

File: modules/core/src/compiler/directive_metadata_reader.js [hunk 5]
```diff
@@ -26,4 +33,16 @@ export class DirectiveMetadataReader {
   parseShadowDomStrategy(annotation:Component):ShadowDomStrategy{
     return isPresent(annotation.shadowDom) ? annotation.shadowDom : ShadowDomNative;
   }
+
+  componentDirectives(annotation:Component, shadowDomStrategy:ShadowDomStrategy):List<DirectiveMetadata> {
```
Comment (rkirov) [inline modules/core/src/compiler/directive_metadata_reader.js]: `componentDirectiveMetadata` is a more descriptive name.

File: modules/core/src/compiler/view.js [hunk 6]
```diff
@@ -38,6 +40,11 @@ export class View {
   proto: ProtoView;
   context: any;
   contextWithLocals:ContextWithVariableBindings;
+  contentTagContainers:List;
+
+  // DEFINE
```
Comment (rkirov) [inline modules/core/src/compiler/view.js]: This comment is not clear to me, remove or expand.

File: modules/core/src/compiler/view.js [hunk 7]
```diff
@@ -319,25 +351,15 @@ export class ProtoView {
...
-      }
-      preBuiltObjects[i] = preBuiltObject;
-
-      // elementsWithPropertyBindings
       if (binder.hasElementPropertyBindings) {
         ListWrapper.push(elementsWithPropertyBindings, element);
       }
 
+      // store a reference to an injector that might have a content tag directive
+      if (this._isContentTag(element)) {
```
Comment (rkirov) [inline modules/core/src/compiler/view.js]: This looks like something that belongs in the compiler. Let's talk whether we can move it.

File: modules/core/src/compiler/element_injector.js [hunk 8]
```diff
@@ -426,10 +434,26 @@ export class ElementInjector extends TreeNode {
...
+      return this._findHostInjector()._preBuiltObjects.lightDom;
+    }
+
     //TODO add other objects as needed
     return _undefined;
   }
 
+  _findHostInjector():ElementInjector {
+    var c = this;
+    while (isBlank(c._host)) {
```
Comment (rkirov) [inline modules/core/src/compiler/element_injector.js]: I thought `_host` is propagated down to children during constuction (see the constructor). So this should not be necessary.

File: modules/core/src/compiler/shadow_dom_emulation/content_tag.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import {Decorator} from '../../annotations/annotations';
+import {SourceLightDom, DestinationLightDom, LightDom} from './light_dom';
+import {Inject} from 'di/di';
+import {Element, Node, DOM} from 'facade/dom';
+import {List, ListWrapper} from 'facade/collection';
+import {NgElement} from 'core/dom/element';
+
+var _scriptTemplate = DOM.createScriptTag("type", "ng/content")
```
Comment (rkirov) [inline modules/core/src/compiler/shadow_dom_emulation/content_tag.js]: petpeeve: use `'` vs `"` consistently.

File: modules/core/src/compiler/shadow_dom_emulation/content_tag.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    this._destinationLightDom = destinationLightDom;
+
+    this.select = contentEl.getAttribute("select");
+
+    this._replaceContentElementWithScriptTags(contentEl.domElement);
+  }
+
+  insert(nodes:List<Node>) {
+    DOM.insertAllBefore(this._endScript, nodes);
+    this._removeNodesUntil(ListWrapper.isEmpty(nodes) ? this._endScript : nodes[0]);
```
Comment (rkirov) [inline modules/core/src/compiler/shadow_dom_emulation/content_tag.js]: Why not remove all between start and end and then just insert? Also is `insert` the wrong name if we remove the previous content.

File: modules/core/src/compiler/shadow_dom_emulation/light_dom.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+function redistributeNodes(contents:List<Content>, nodes:List<Node>) {
+  for (var i = 0; i < contents.length; ++i) {
+    var content = contents[i];
+    var select = content.select;
+    var matchSelector = (n) => DOM.elementMatches(n, select);
+
+    if (isBlank(select)) {
+      content.insert(nodes);
+      return;
```
Comment (rkirov) [inline modules/core/src/compiler/shadow_dom_emulation/light_dom.js]: no need for early return

File: modules/core/src/compiler/shadow_dom_emulation/light_dom.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+export class LightDom {
+  lightDomView:View;
+  shadowDomView:View;
+  element:Element;
+  roots:List<Node>;
+
+  constructor(lightDomView:View, shadowDomView:View, element:Element) {
+    this.lightDomView = lightDomView;
+    this.shadowDomView = shadowDomView;
+    this.element = element;
```
Comment (rkirov) [inline modules/core/src/compiler/shadow_dom_emulation/light_dom.js]: do we need to keep this.element? Why not just pop the nodes and pass `this.roots`?

File: modules/core/src/compiler/view.js [hunk 13]
```diff
@@ -191,6 +205,17 @@ export class View {
     }
   }
 
+  getViewPortByTemplateElement(node):ViewPort {
```
Comment (rkirov) [inline modules/core/src/compiler/view.js]: don't fully understand this, but I think we should be able to get this from the `elementInjectors`

File: modules/core/src/compiler/element_injector.js (hunk 8 above)
Comment (vsavkin) [inline modules/core/src/compiler/element_injector.js]: ElementInjector has either a host or a parent.

File: modules/core/test/compiler/integration_spec.js [hunk 14]
```diff
@@ -108,9 +109,51 @@ export function main() {
         });
       });
     });
+
+    it('should emulate content tag', (done) => {
+      var el = `<emulated-shadow-dom-component>` +
+        `<div>Light</div>` +
+        `<div template="trivial-template">DOM</div>` +
+      `</<emulated-shadow-dom-component>`;
```
Comment (rkirov) [inline modules/core/test/compiler/integration_spec.js]: typo `</<`.

File: modules/facade/src/dom.dart [hunk 15]
```diff
@@ -95,9 +110,15 @@ class DOM {
   static hasClass(Element element, classname) {
     return element.classes.contains(classname);
   }
+  static String tagName(Element element) {
```
Comment (rkirov) [inline modules/facade/src/dom.dart]: Looks the same in dart and js. Why is this needed? Same for getAttribute.

File: modules/core/test/compiler/directive_metadata_reader_spec.js [hunk 16]
```diff
@@ -17,13 +29,28 @@ class ComponentWithoutExplicitShadowDomStrategy {}
...
 class SomeDirectiveWithoutAnnotation {
 }
 
+@Component({
+  selector: 'withoutDirectives'
+})
+class ComponentWithoutDirectives {}
+
+@Component({
+  selector: 'withoutDirectives',
```
Comment (vicb) [inline modules/core/test/compiler/directive_metadata_reader_spec.js]: 'withDirectives' ?

File: modules/core/src/compiler/element_injector.js (hunk 8 above)
Comment (vicb) [inline modules/core/src/compiler/element_injector.js]: I have the same understanding as Rado here

File: modules/core/src/compiler/shadow_dom_emulation/content_tag.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  _replaceContentElementWithScriptTags(contentEl:Element) {
+    this._beginScript = DOM.clone(_scriptTemplate);
+    this._endScript = DOM.clone(_scriptTemplate);
+
+    DOM.insertBefore(contentEl, this._beginScript);
```
Comment (vicb) [inline modules/core/src/compiler/shadow_dom_emulation/content_tag.js]: should it be `contentEl.domElement` ?

File: modules/core/test/compiler/element_injector_spec.js [hunk 18]
```diff
@@ -88,12 +93,15 @@ export function main() {
     return inj;
   }
 
-  function parentChildInjectors(parentBindings, childBindings) {
+  function parentChildInjectors(parentBindings, childBindings, parentPreBuildObjects = null) {
```
Comment (vicb) [inline modules/core/test/compiler/element_injector_spec.js]: Why not default to defaultPreBuiltObject ?

File: modules/facade/src/dom.dart (hunk 15 above)
Comment (vsavkin) [inline modules/facade/src/dom.dart]: I thought the plan is always go through the facade for DOM, so we can instrument it.

File: modules/core/test/compiler/element_injector_spec.js (hunk 18 above)
Comment (vsavkin) [inline modules/core/test/compiler/element_injector_spec.js]: It is not a compile-time constant. I don't think we have a way right not to declare compile-time constants.

File: modules/core/src/compiler/view.js (hunk 13 above)
Comment (vsavkin) [inline modules/core/src/compiler/view.js]: This is still needed for expanding view ports.

Comment (rkirov) [inline modules/core/src/compiler/view.js]: Fair, add a todo to see if this can be done only once during LightDom creation, instead of during distribution.

Comment (vsavkin) [inline modules/core/src/compiler/view.js]: Done.
